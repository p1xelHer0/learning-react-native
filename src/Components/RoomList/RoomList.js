import React from 'react'
import { ListView, StyleSheet } from 'react-native'

import Room from '../Room/Room'

const { arrayOf, func, object } = React.PropTypes

const propTypes = {
  POINT_APPROVE: func.isRequired,
  POINT_COMMENT_SET: func.isRequired,
  POINT_DECLINE: func.isRequired,
  POINT_TOGGLE_APPROVAL: func.isRequired,
  POINT_TOGGLE_DETAILS: func,
  ROOM_POINTS_TOGGLE: func,
  points: arrayOf(object),
  rooms: arrayOf(object)
}

const RoomList = ({
  POINT_APPROVE,
  POINT_COMMENT_SET,
  POINT_DECLINE,
  POINT_TOGGLE_APPROVAL,
  POINT_TOGGLE_DETAILS,
  ROOM_POINTS_TOGGLE,
  points,
  rooms
}) => {
  const ds = new ListView.DataSource({
    rowHasChanged: (row1, row2) => row1 !== row2
  })
  const dataSource = ds.cloneWithRows(rooms)
  return (
    <ListView
      enableEmptySections
      style={localStyles.container}
      dataSource={dataSource}
      renderRow={
        room =>
          <Room
            {...room}
            POINT_APPROVE={POINT_APPROVE}
            POINT_COMMENT_SET={POINT_COMMENT_SET}
            POINT_DECLINE={POINT_DECLINE}
            POINT_TOGGLE_APPROVAL={POINT_TOGGLE_APPROVAL}
            POINT_TOGGLE_DETAILS={POINT_TOGGLE_DETAILS}
            key={room.id}
            points={points}
            togglePoints={() => ROOM_POINTS_TOGGLE(room.id)}
          />
      }
    />
  )
}

const localStyles = StyleSheet.create({
  container: {
    flex: 1
  }
})

RoomList.propTypes = propTypes

export default RoomList
