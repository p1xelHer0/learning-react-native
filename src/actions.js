// FSA-compliant actions
// see https://github.com/acdlite/flux-standard-action

// searchReducer
export const SEARCH_TERM_SET = searchTerm => ({
  type: 'SEARCH_TERM_SET',
  payload: { searchTerm }
})

export const SEARCH_TERM_REMOVE = () => ({
  type: 'SEARCH_TERM_REMOVE'
})

// roomReducer
export const ROOM_ADD = (location, roomId) => ({
  type: 'ROOM_ADD',
  payload: {
    roomId,
    room: { id: roomId, location, pointIds: [] }
  }
})

export const ROOM_POINT_ADD = (pointId, roomId) => ({
  type: 'ROOM_POINT_ADD',
  payload: {
    roomId,
    pointId
  }
})

export const ROOM_POINTS_TOGGLE = roomId => ({
  type: 'ROOM_POINTS_TOGGLE',
  payload: { roomId }
})

// pointReducer
export const POINT_ADD = (point, pointId) => ({
  type: 'POINT_ADD',
  payload: { pointId, point }
})

export const POINT_APPROVE = pointId => ({
  type: 'POINT_APPROVE',
  payload: { pointId }
})

export const POINT_DECLINE = pointId => ({
  type: 'POINT_DECLINE',
  payload: { pointId }
})

export const POINT_COMMENT_SET = (comment, pointId) => ({
  type: 'POINT_COMMENT_SET',
  payload: {
    pointId,
    comment
  }
})

export const POINT_TOGGLE_APPROVAL = pointId => ({
  type: 'POINT_TOGGLE_APPROVAL',
  payload: { pointId }
})

export const POINT_TOGGLE_DETAILS = pointId => ({
  type: 'POINT_TOGGLE_DETAILS',
  payload: { pointId }
})

// categoryReducer
export const CATEGORY_ADD = (categoryName, categoryId) => ({
  type: 'CATEGORY_ADD',
  payload: { categoryId, categoryName }
})
