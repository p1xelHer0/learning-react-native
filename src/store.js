import { applyMiddleware, combineReducers, compose, createStore } from 'redux'
import { connect } from 'react-redux'
import thunk from 'redux-thunk'
import { createSelector } from 'reselect'

import * as actions from './actions'

import categoryReducer, * as fromCategories from './Reducers/categories'
import pointReducer, * as fromPoints from './Reducers/points'
import roomReducer, * as fromRooms from './Reducers/rooms'
import searchTermReducer, * as fromSearchTerm from './Reducers/searchTerm'

export const initialState = {
  categoriesById: fromCategories.initialState,
  pointsById: fromPoints.initialState,
  roomsById: fromRooms.initialState,
  searchTerm: fromSearchTerm.initialState
}

export const rootReducer = combineReducers({
  categoriesById: categoryReducer,
  pointsById: pointReducer,
  roomsById: roomReducer,
  searchTerm: searchTermReducer
})

// enable reactNativeDebugger
// see https://github.com/jhen0409/react-native-debugger
const reactNativeDebuggerHelper =
  global.reduxNativeDevTools ? global.reduxNativeDevTools(/* options */) : noop => noop

// thunk is needed for React Native Debugger
const middleware = compose(
  applyMiddleware(thunk),
  reactNativeDebuggerHelper
)

export const store = createStore(rootReducer, initialState, middleware)

// reduxNativeDevTools
if (global.reduxNativeDevTools) {
  global.reduxNativeDevTools.updateStore(store)
}

// helpers for memoized selectors
const getCategoriesById = state => state.categoriesById
const getPointsById = state => state.pointsById
const getRoomsById = state => state.roomsById
const getSearchTerm = state => state.searchTerm

// memoized selectors

// filter the rooms by the searchTerm from the store
// the implementation itself is in the room reducer (rooms.js)
export const getFilteredRoomsSelector = createSelector(
  [getRoomsById, getSearchTerm],
  (getRoomsById, getSearchTerm) => getFilteredRooms(getRoomsById, getSearchTerm)
)

// call the implementation
export const getFilteredRooms = (roomsById = {}, filter = '') =>
  fromRooms.getFilteredRooms(roomsById, filter)

// selector which sets the category name on the point property
// this way, we don't have to refer to the catagoriesById in the
// store to get the name of the category
// the implementation itself is in the points reducer (points.js)
export const getPointsWithCategorySelector = createSelector(
  [getPointsById, getCategoriesById],
  (getPointsById, getCategoriesById) => getPointsWithCategory(getPointsById, getCategoriesById)
)

// call the implementation
export const getPointsWithCategory = (pointsById = {}, categoriesById = {}) =>
  fromPoints.getPointsWithCategory(pointsById, categoriesById)

const mapStateToProps = state => ({
  categoriesById: state.categoriesById,
  points: getPointsWithCategorySelector(state),
  rooms: getFilteredRoomsSelector(state),
  searchTerm: state.searchTerm
})

export const connector = connect(mapStateToProps, { ...actions })
