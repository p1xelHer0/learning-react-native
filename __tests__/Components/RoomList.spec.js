/* eslint-env jest */
import React from 'react'
import 'react-native'

// test renderer must be required after react-native
import renderer from 'react-test-renderer'

import { initialState, getFilteredRooms } from '../../src/store'
import RoomList from '../../src/Components/RoomList/RoomList'

// we don't want the test to depend on styling
const testStyle = {}

describe('<RoomList />', () => {
  it('should match the snapshot', () => {
    const rooms = getFilteredRooms(initialState.roomsById, initialState.searchTerm)
    const roomList = renderer.create(
      <RoomList rooms={rooms} style={testStyle} />
    ).toJSON()
    expect(roomList).toMatchSnapshot()
  })
})
