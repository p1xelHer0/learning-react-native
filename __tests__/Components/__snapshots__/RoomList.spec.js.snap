exports[`<RoomList /> should match the snapshot 1`] = `
<ScrollView
  dataSource={
    ListViewDataSource {
      "items": 5,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={true}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "flex": 1,
    }
  }>
  <View>
    <View
      style={
        Object {
          "borderColor": "#433d35",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
        }
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "flex": 1,
            "padding": 5,
          }
        }>
        Entré
      </Text>
      <View
        style={
          Object {
            "padding": 5,
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail">
          12
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            undefined,
          ]
        }
        testID={undefined}>
        <View
          style={
            Array [
              Object {
                "height": 35,
                "width": 35,
              },
              Object {
                "backgroundColor": "#60584c",
              },
              undefined,
            ]
          } />
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        Object {
          "borderColor": "#433d35",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
        }
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "flex": 1,
            "padding": 5,
          }
        }>
        Orange samtal
      </Text>
      <View
        style={
          Object {
            "padding": 5,
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail">
          2
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            undefined,
          ]
        }
        testID={undefined}>
        <View
          style={
            Array [
              Object {
                "height": 35,
                "width": 35,
              },
              Object {
                "backgroundColor": "#60584c",
              },
              undefined,
            ]
          } />
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        Object {
          "borderColor": "#433d35",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
        }
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "flex": 1,
            "padding": 5,
          }
        }>
        Rosa samtal
      </Text>
      <View
        style={
          Object {
            "padding": 5,
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail">
          2
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            undefined,
          ]
        }
        testID={undefined}>
        <View
          style={
            Array [
              Object {
                "height": 35,
                "width": 35,
              },
              Object {
                "backgroundColor": "#60584c",
              },
              undefined,
            ]
          } />
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        Object {
          "borderColor": "#433d35",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
        }
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "flex": 1,
            "padding": 5,
          }
        }>
        Röd konferens
      </Text>
      <View
        style={
          Object {
            "padding": 5,
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail">
          2
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            undefined,
          ]
        }
        testID={undefined}>
        <View
          style={
            Array [
              Object {
                "height": 35,
                "width": 35,
              },
              Object {
                "backgroundColor": "#60584c",
              },
              undefined,
            ]
          } />
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        Object {
          "borderColor": "#433d35",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
        }
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "flex": 1,
            "padding": 5,
          }
        }>
        Pentry
      </Text>
      <View
        style={
          Object {
            "padding": 5,
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail">
          5
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            undefined,
          ]
        }
        testID={undefined}>
        <View
          style={
            Array [
              Object {
                "height": 35,
                "width": 35,
              },
              Object {
                "backgroundColor": "#60584c",
              },
              undefined,
            ]
          } />
      </View>
    </View>
  </View>
</ScrollView>
`;
